fs/btrfs/zstd.c

```diff
@@ -95,7 +114,26 @@ static int zstd_compress_pages(struct list_head *ws,
 	unsigned long len = *total_out;
 	const unsigned long nr_dest_pages = *out_pages;
 	unsigned long max_out = nr_dest_pages * PAGE_SIZE;
-	ZSTD_parameters params = zstd_get_btrfs_parameters(len);
+
+	ZSTD_parameters params = zstd_get_btrfs_parameters(len, workspace->level);
+	size_t new_size = ZSTD_CStreamWorkspaceBound(params.cParams);
+	if (new_size > workspace->size) {
```

Comment:
I'm not crazy about this silent fallback. If I ask for the max compression level, then I expect it to really use that compression level, and I'd like it to fail hard otherwise. If you change `zstd_alloc_workspace()` like I mentioned above, then this is only an issue for remount. Maybe instead we could do the reallocation at remount-time, and we could fail the remount if we can't? Something like preload with a workspace of the desired level and try to resize or throw away everything else? How much bigger is each level?

Reply:
I think we should be able to change `alloc_workspace()` to accept a compression level, and we can hard fail there. The only caveat is that I'm not sure if we have the level in `btrfs_init_compress()`. If we can't get that info, we can always preallocate the first workspace with the highest level.

IIRC the memory usage ranges from 1 MB to a few MB. @jennifermliu can get you the exact numbers.

Does this sound like a reasonable direction?
1. Pass the compression level to `alloc_workspace()`.
2. `set_level()` either sets the compression level if there is enough memory available, or fails.
3. Preallocate the first workspace with the highest level. This way we keep the forward progress guarantees.
4. Call `set_level()` in `find_workspace()` before succeeding. If it fails, free the workspace and attempt to allocate a new one with the new memory usage.
5. Now when we get a workspace, we know we have enough memory available.

For step 3, we can either preallocate one with the highest level, or at remount time reallocate and fail if we are out of memory. Either works for me, so whatever you think is better. If we go with the latter case, we'll have to move the allocation logic out of `btrfs_init_compress()`, since we don't have the level there.

Reply:
CLevel | Memory (in bytes)
------------ | -------------
1 | 799376
2 | 1028752
3 | 1290896
4 | 930448
5 | 1421968
6 | 1553040
7 | 1421968
8 | 1815184
9 | 1815184
10 | 1815184
11 | 1815184
12 | 1815184
13 | 2339472
14 | 1815184
15 | 1815184

Reply:
You're right, `btrfs_init_compress()` is called at kernel init/module load time. I'd be fine with allocating one of the max size there. The rest of your plan sounds sane to me, thanks!